src/embed_tests/Codecs.cs

```diff
@@ -57,30 +66,188 @@ public class Codecs {
         }
 
         [Test]
-        public void TupleRoundtripObject() {
+        public void TupleRoundtripObject()
+        {
             TupleRoundtripObject<ValueTuple<int, string, object>, ValueTuple>();
         }
-        static void TupleRoundtripObject<T, TTuple>() {
+        static void TupleRoundtripObject<T, TTuple>()
+        {
             var tuple = Activator.CreateInstance(typeof(T), 42, "42", new object());
-            using (Py.GIL()) {
+            using (Py.GIL())
+            {
                 var pyTuple = TupleCodec<TTuple>.Instance.TryEncode(tuple);
                 Assert.IsTrue(TupleCodec<TTuple>.Instance.TryDecode(pyTuple, out object restored));
                 Assert.AreEqual(expected: tuple, actual: restored);
             }
         }
 
         [Test]
-        public void TupleRoundtripGeneric() {
+        public void TupleRoundtripGeneric()
+        {
             TupleRoundtripGeneric<ValueTuple<int, string, object>, ValueTuple>();
         }
 
-        static void TupleRoundtripGeneric<T, TTuple>() {
+        static void TupleRoundtripGeneric<T, TTuple>()
+        {
             var tuple = Activator.CreateInstance(typeof(T), 42, "42", new object());
-            using (Py.GIL()) {
+            using (Py.GIL())
+            {
                 var pyTuple = TupleCodec<TTuple>.Instance.TryEncode(tuple);
                 Assert.IsTrue(TupleCodec<TTuple>.Instance.TryDecode(pyTuple, out T restored));
                 Assert.AreEqual(expected: tuple, actual: restored);
             }
         }
+
+        [Test]
+        public void FunctionAction()
+        {
+            FunctionCodec.Register();
+            var codec = FunctionCodec.Instance;
+
+            //decoding - python functions to C# actions
+            {
+                PyInt x = new PyInt(1);
+                PyDict y = new PyDict();
+                //non-callables can't be decoded into Action
+                Assert.IsFalse(codec.CanDecode(x, typeof(Action)));
+                Assert.IsFalse(codec.CanDecode(y, typeof(Action)));
+
+                var locals = new PyDict();
+                PythonEngine.Exec(@"
+def foo():
+    return 1
+def bar(a):
+    return 2
+", null, locals.Handle);
+
+                //foo, the function with no arguments
+                var fooFunc = locals.GetItem("foo");
+                Assert.IsFalse(codec.CanDecode(fooFunc, typeof(bool)));
+
+                //CanDecode does not work for variadic actions
+                //Assert.IsFalse(codec.CanDecode(fooFunc, typeof(Action<object[]>)));
```

Comment:
@lostmsu because of the interface of CanDecode this doesn't work.  Since I defined functions in C# to use object[] I cannot determine the number of arguments without an instance.

Reply:
You can create a special delegate type just for variadic parameter handling, and check for it explicitly in the codec.
E.g. `delegate void VariadicAction<T>(params T[] args)`

Reply:
@lostmsu The issue is with the number of arguments, not the type of the arguments

Reply:
@lostmsu I think I get your point.  Do you mean rather than Action<object[]> and Func<object[], object> I can simply use VariadicAction so if I have a situation like this:

C#
public int CallMe(Func<int, int, string, double, object> func) {
    return CallMe(1,2, "hello", 4.4, new SomeObject());
}

Python
def call_me(i, j, name, real_value, obj):
    return 1

You except that given typeof(func) you can convert it to VariadicAction and call Invoke on it dynamically?  I can try and I think this may work except it may have a restriction to use object in all of the argument slots.